framework/src/modules/interoperability/mainchain/module.ts

```diff
@@ -367,32 +367,44 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
 						'For each chainInfo with status terminated there should be a corresponding entry in terminatedStateAccounts.',
 					);
 				}
+			}
+		}
 
```

Comment:
I think current implementation is correct and according to the LIP.

However, I think that in terms of performance, we are introducing way many un-necessary loops. 

Perhaps it would have been better if we just move the call to `				this._verifyTerminatedStateAccountsCommon(terminatedStateAccounts, mainchainID);` in between the first two loops (e.g., line 357) and then leave the second for loop as it was before. 

The reason is that in the second for-loop (line 358 onwards) we loop through chainInfos and find the corresponding terminated state account, so all checks can be done at this step. 

By closing the loop and starting a new loop in line 381 all those pairs of entries in chainInfos and terminated state accounts have to be found again, which is a bit wasteful.

Reply:
In fact even calling the `verifyTerminatedStateAccountsCommon` is not that efficient since it introduces an un-necessary loop, but we did it for avoiding repetition and errors, so it is fine to keep it I guess..

Reply:
In fact I think here is the most efficient way to do this, minimizing the number of for-loops (@ishantiw we can discuss this all together to see if we proceed):

- Remove the for-loop of lines 358-371 and perform this check in `verifyChainData`, which runs inside a loop across `chainInfos` anyways. In other words, move lines 361-370 after line 334 (with appropriate modification of variable names). This way we check the one direction of "if-and-only-if", i.e, that each chain with status terminated should have a corresponding terminated account.
- Then, during `verifyTerminatedStateAccounts`, loop *only once* through the `terminatedStateAccounts` array, as follows:  

1.  Check uniqueness of `chainID` and sorting in lexicographic order (now lines 196-211 in `_verifyTerminatedStateAccountsCommon`).
2. Loop through elements `stateAccount` of `terminatedStateAccounts` and check:2a) that there exists a corresponding entry in `chainInfo`, otherwise throw (lines 345-355 now) 2b) `verifyChainID(stateAccount.chainID)`, (now line 213 of `_verifyTerminatedStateAccountsCommon`) and then 2c) verify the validity of `stateAccount.terminatedStateAccount` as in lines 388-408.

Note this would have the exact same result, and replaces 4 loops by just 1. The only refactoring is that we don't need anymore `_verifyTerminatedStateAccountsCommon`, so we need to adjust the sidechain check as well.